src/api.py

```diff
@@ -0,0 +1,146 @@
+import os
+
+from fastapi import FastAPI
+from fastapi.responses import RedirectResponse, FileResponse
+from fastapi.staticfiles import StaticFiles
+
+from .charset import URLCharset
+from .codec import Codec
+from .database import ViteDB
+
+## CONSTANTS ##
+
+PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+STATIC_PATH  = os.path.join(PROJECT_ROOT, 'src', 'static')
+DB_PATH      = os.path.join(PROJECT_ROOT, 'data', 'vite.db')
+
+DOMAIN_NAME  = "http://vite.lol/"
+SHORT_URL    = DOMAIN_NAME[7:] # without the http://
+
+## CORE LOGIC ##
+
+url_charset = URLCharset(numeric=True, lowercase_ascii=True,
+                         uppercase_ascii=True, special=False)
+codec       = Codec(charset=url_charset)
+app         = FastAPI(docs_url="/docs/")
+
+# Create the database if it doesn't exist and the links table
+os.makedirs(os.path.dirname(DB_PATH), exist_ok=True)
+with ViteDB(DB_PATH) as db:
+    db.create_table(db.table_name, db.fields)
+    
+# "/static/" avoids collisions with a possible /static generated path in the future
+app.mount("/static/", StaticFiles(directory=STATIC_PATH), name="static")
+    
+## API ENDPOINTS ##
+
+@app.get("/")
+def read_root() -> dict:
+    return FileResponse(f"{STATIC_PATH}/index.html")
+
+@app.get("/encode")
+def encode_url(url: str) -> dict:
+    """Encodes an URL to a short string
+
+    Args:
+        url (str): The URL to encode
+
+    Returns:
+        dict: A JSON response containing the encoded URL
+    """
+        
+    if url == "":
+        return {"error": "No URL provided"}
+    
+    with ViteDB(DB_PATH) as db:
+        row_count = db.get_row_count()
+        
+    unique_id: int = row_count + 1
+    encoded_uid: str = codec.encode(unique_id)
+    
+    shortened_url: str = f"{DOMAIN_NAME}{encoded_uid}"
+    
+    with ViteDB(DB_PATH) as db:
+        db.insert_link(url)
+    
+    return {"url": shortened_url}
+
+
+@app.get("/decode")
+def decode_url(url: str) -> dict:
+    """Decodes a short string to its original URL
+
+    Args:
+        url (str): The URL to decode
+
+    Returns:
+        dict: A JSON response containing the original URL and the number of clicks
+    """
+        
+    if url.startswith(DOMAIN_NAME):
+        url = url[len(DOMAIN_NAME):]
+    elif url.startswith(SHORT_URL):
+        url = url[len(SHORT_URL):]
+        
+    if url == "":
+        return {"error": "No URL provided"}
+    elif url == "0": # There's no row id 0, so there can't be a shortened URL for it
+        return {"url": "https://en.wikipedia.org/wiki/0#Computer_science", "clicks": -1}
+    elif url_charset.validate(url) == False:
+        return {"error": "Not a valid URL"}
```

Comment:
these checks need to be performed earlier because the previous test uses the url parameter, so if it's null, you'll never get to your test

Reply:
Sorry, I don't think I understand what you mean, if it's null they get the standard pydantic return protocol as such:

```json
{
  "detail": [
    {
      "type": "missing",
      "loc": [
        "query",
        "value"
      ],
      "msg": "Field required",
      "input": null,
      "url": "https://errors.pydantic.dev/2.6/v/missing"
    }
  ]
}```

I assumed this was clear enough and didn't need re-implementation, but maybe I'm misunderstanding what you mean, could you provide a usage example that triggers a bug?

I got this from calling:

`/decode` and nothing else